docs/puzzles/day6.md

```diff
@@ -2,12 +2,336 @@
+  //
+  // This list means that the first fish has an internal timer of 3, the second fish
+  // has an internal timer of 4, and so on until the fifth fish, which has an
+  // internal timer of 2."
+  def parseSeveral(input: String): Seq[Fish] =
+    for timerString <- input.trim.split(",").toIndexedSeq
+    yield Fish(timerString.toInt.ensuring(_ >= 0))
+
+/**
+ * Simulate the evolution of the population and return the number
+ * of fishes at the end of the simulation.
+ * @param days Number of days to simulate
+ * @param initialPopulation Initial population
+ */
+def simulate(days: Int, initialPopulation: Seq[Fish]): Int =
+  (1 to days)
+    .foldLeft(initialPopulation)((population, _) => tick(population))
+    .size
+
+/**
+ * Compute a new population after one day passes.
+ * @param population Current population
+ * @return New population
+ */
+def tick(population: Seq[Fish]): Seq[Fish] =
+  population.flatMap { fish =>
+    // "Each day, a `0` becomes a `6` and adds a new `8` to the end of the list"
+    if fish.timer == 0 then
+      Seq(Fish(6), Fish(8))
+    // "while each other number decreases by 1"
+    else
+      Seq(Fish(fish.timer - 1))
+  }
+~~~
+
+<Solver puzzle="day6-part1"/>
+
+## First attempt for Part 2
+
+The challenge for the part 2 does not seem complex at the first sight:
+
+> How many lanternfish would there be after 256 days?
+
+Let’s just run our simulation for 256 days instead of 80 days, and we are good!
+
+~~~ scala
+def part2(input: String): Int =
+  simulate(
+    days = 256,
+    initialPopulation = Fish.parseSeveral(input)
+  )
+~~~
+
+Unfortunately, after running for a few minutes, it crashed with the 
+following error:
+
+~~~ text
+Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
+        at scala.reflect.ManifestFactory$AnyManifest.newArray(Manifest.scala:328)
+        at scala.reflect.ManifestFactory$AnyManifest.newArray(Manifest.scala:327)
+        at scala.collection.IterableOnceOps.toArray(IterableOnce.scala:1278)
+        at scala.collection.IterableOnceOps.toArray$(IterableOnce.scala:1276)
+        at scala.collection.AbstractIterable.toArray(Iterable.scala:919)
+        at scala.collection.immutable.ArraySeq$.$anonfun$newBuilder$1(ArraySeq.scala:278)
+        at scala.collection.immutable.ArraySeq$$$Lambda$28/0x000000084010c040.apply(Unknown Source)
+        at scala.collection.mutable.Builder$$anon$1.result(Builder.scala:83)
+        at scala.collection.StrictOptimizedIterableOps.flatMap(StrictOptimizedIterableOps.scala:119)
+        at scala.collection.StrictOptimizedIterableOps.flatMap$(StrictOptimizedIterableOps.scala:104)
+        at scala.collection.immutable.ArraySeq.flatMap(ArraySeq.scala:35)
+        at day6$package$.tick(day6.scala:39)
+        at day6$package$.simulate$$anonfun$1(day6.scala:29)
+        at day6$package$.simulate$$anonfun$adapted$1(day6.scala:29)
+~~~
+
+Oops.
+
+What happens is that there are too many fish, and computing the collection 
```

Comment:
```suggestion
What happens is that there are too many fishes, and computing the collection 
```

Reply:
I believe that fish is correct in this example

Reply:
Oh I just learned that the plural of fish could be fish.